numba/tests/test_array_methods.py

```diff
@@ -1329,6 +1335,45 @@ def test_imag(self):
         z = x + 1j*y
         np.testing.assert_equal(pyfunc(z), cfunc(z))
 
+    def test_clip(self):
+        # TODO: scalars are not tested (issue #3469)
+        for a in (np.linspace(-10, 10, 101),
+                  np.linspace(-10, 10, 40).reshape(5, 2, 4)):
+            for pyfunc in [np_clip, array_clip]:
+                cfunc = jit(nopython=True)(pyfunc)
+
+                with self.assertRaises(ValueError):
+                    pyfunc(a, None, None)
+                with self.assertRaises(ValueError):
+                    cfunc(a, None, None)
+
+                np.testing.assert_equal(pyfunc(a, 0, None), cfunc(a, 0, None))
+                np.testing.assert_equal(pyfunc(a, None, 0), cfunc(a, None, 0))
+
+                np.testing.assert_equal(pyfunc(a, -5, 5), cfunc(a, -5, 5))
+
+                pyout = np.empty_like(a)
+                cout = np.empty_like(a)
+                np.testing.assert_equal(pyfunc(a, -5, 5, pyout),
+                                        cfunc(a, -5, 5, cout))
+                np.testing.assert_equal(pyout, cout)
+
+                # This test exhibits undefined behaviour since min > max, but
+                # we'll make sure it's consistent with NumPy, which computes
+                # min(max(val, a_min), a_max) since 1.17. Prior to that min and
+                # max were the other way round, so we're not consistent with
+                # older NumPy versions.
+                if numpy_version > (1, 16):
+                    np.testing.assert_equal(pyfunc(a, 8, 5), cfunc(a, 8, 5))
+
+                # verifies that type-inference is working on the return value
+                # this used to trigger issue #3489
+                def lower_clip_result(a):
+                    return np.expm1(cfunc(a, -5, 5))
+                np.testing.assert_almost_equal(
+                    lower_clip_result(a),
+                    jit(nopython=True)(lower_clip_result)(a))
```

Comment:
I'm not sure what this is about in relation to #3489, any chance some more explanation could be offered please?

Reply:
From https://github.com/numba/numba/pull/3468#issuecomment-439137556, it looks like this is testing the case where the returned value is a kwarg, which #3489 reports as not working.

Reply:
Thanks.